### PR TITLE
Teach the importer to import any clang vector type as SIMDn<Scalar>.

### PR DESCRIPTION
**Pull request by stephentyrone**
Instead of having a known set of vector types, check to see if the element type conforms to SIMDScalar; if it does, see if we have a SIMDn defined with the right number of elements. If both are satisfied, import the vector type as that Swift type.

By making this change, we gain the ability to import vector types that aren't defined in terms of the Darwin simd module, which lets us use C API with vector types on other platforms. It also lets us
import *every* vector type that Swift can represent, rather than the small subset that are currently hardcoded.

Resovles <rdar://problem/34170174>

### CODE REVIEW

**Comment by stephentyrone** [comment]
@rjmccall I'm probably doing a whole bunch of things wrong here, and I still need to implement the reverse mapping, but this is functional enough to get some feedback on while I work on the rest.

**Comment by stephentyrone** [comment]
@swift-ci please smoke test

**Comment by rjmccall** [comment]
Are these different types from what we use to import these as?

File: lib/ClangImporter/ImportType.cpp [hunk 1]
```diff
@@ -510,36 +510,32 @@ namespace {
...
+      auto nominal = element->getAnyNominal();
+      auto simdscalar = Impl.SwiftContext.getProtocol(KnownProtocolKind::SIMDScalar);
+      SmallVector<ProtocolConformance *, 2> conformances;
+      if (simdscalar && nominal->lookupConformance(nominal->getParentModule(),
+                                                   simdscalar, conformances)) {
+        SmallString<8> name("SIMD");
+        name.append(std::to_string(count));
+        if (auto vector = Impl.getNamedSwiftType(Impl.getStdlibModule(), name)) {
+          if (auto unbound = vector->getAs<UnboundGenericType>()) {
+            ArrayRef<Type> arg(element);
```
Comment (rjmccall) [inline lib/ClangImporter/ImportType.cpp]: I'm not sure this is okay memory-wise, but you should be able to just pass `{ arg }`.

File: lib/ClangImporter/ImportType.cpp [hunk 2]
```diff
@@ -510,36 +510,32 @@ namespace {
...
-            return Type();
+      // Get the element type and count.
+      Type element = Impl.importTypeIgnoreIUO(type->getElementType(),
+                                              ImportTypeKind::Abstract,
+                                              false,
+                                              Bridgeability::Full,
+                                              OptionalTypeKind::OTK_None);
+      unsigned count = type->getNumElements();
+      // Import vector-of-one as the element type.
+      if (count == 1) { return element; }
```
Comment (rjmccall) [inline lib/ClangImporter/ImportType.cpp]: Hmm.  Does this always have the same layout as it would in C?

File: lib/ClangImporter/ImportType.cpp [hunk 3]
```diff
@@ -510,36 +510,32 @@ namespace {
...
-#include "swift/ClangImporter/SIMDMappedTypes.def"
-          default:
-            // A vector type we don't know how to map.
-            return Type();
+      // Get the element type and count.
+      Type element = Impl.importTypeIgnoreIUO(type->getElementType(),
+                                              ImportTypeKind::Abstract,
+                                              false,
+                                              Bridgeability::Full,
+                                              OptionalTypeKind::OTK_None);
```
Comment (rjmccall) [inline lib/ClangImporter/ImportType.cpp]: I think you want `Bridgeability::None` here, although it probably doesn't actually matter.

Please add a `/* ... */` comment explaining the `false` argument.

**Comment by stephentyrone** [comment]
@rjmccall we used to implement them as types that are typedefs of these types.

File: lib/ClangImporter/ImportType.cpp (hunk 2 above)
Comment (stephentyrone) [reply lib/ClangImporter/ImportType.cpp]: Up to manually-specified alignment, I think so. (How do we handle int32 typedefs that are over/under-aligned for that matter?)

**Comment by rjmccall** [comment]
Okay, that shouldn't cause source-compatibility problems, then.

File: lib/ClangImporter/ImportType.cpp (hunk 2 above)
Comment (rjmccall) [reply lib/ClangImporter/ImportType.cpp]: We probably ignore the alignment, which is probably a bug. :)

Comment (stephentyrone) [reply lib/ClangImporter/ImportType.cpp]: As long as it's consistently a bug, the same as other imported types =)

**Comment by stephentyrone** [comment]
@swift-ci please test

**Comment by stephentyrone** [comment]
@swift-ci please test source compatibility

**Comment by stephentyrone** [comment]
@rjmccall Are you OK with me merging this assuming the compatibility test results are not unexpected?

**Comment by rjmccall** [comment]
Yeah, go ahead and merge.

**Pull request merged**